update `can-read?` and `can-write?` for Field

johnswanson opened the pull request:
The previous behavior was to cache the permissions set for the Field's table. I split this in two: we cache the `db_id` for a given `Field`, but we can tackle caching the specific permissions for that database down the road, when we implement caching on data permissions checks in general.

To fix tests: `set-table-permissions!` previously threw an exception if passed an empty map of `table->permission-value`. I modified it to just skip processing in this case instead.

johnswanson committed:
update `can-read?` and `can-write?` for Field

The previous behavior was to cache the permissions set for the Field's
table. I split this in two: we cache the `db_id` for a given `Field`,
but we can tackle caching the specific permissions for that database
down the road, when we implement caching on data permissions checks in
general.

johnswanson committed:
`set-table-permissions!` works with empty perms

`set-table-permissions!` previously threw an exception if passed an
empty map of `table->permission-value`.

johnswanson committed:
Fix Field `can-read` permission

I think I was wrong about what permission was needed here - to read a
Field, we need `unrestricted` access, not `no-self-service`.

I also updated a failing test to set `DataPermissions` as well as
`Permissions`.

johnswanson committed:
oops: restore data permissions in some tests

johnswanson committed:
test fix: set data-perms too when adjusting perms

johnswanson committed:
fix select-condition for permission test util

johnswanson committed:
fix embedding permissions

johnswanson committed:
fix dashboard tests

johnswanson committed:
fix table tests

johnswanson committed:
lint

johnswanson committed:
only delete/restore sandboxes in EE

escherize commented on test/metabase/api/automagic_dashboards_test.clj:
```diff
@@ -49,24 +49,19 @@
-                      (try
-                        (testing "Endpoint should return 403 if user does not have permissions"
-                          (data-perms/set-database-permission! (perms-group/all-users) (mt/id) :perms/data-access :no-self-service)
-                          (perms/revoke-data-perms! (perms-group/all-users) (mt/id))
+       (let [api-endpoint (apply format (str "automagic-dashboards/" template) args)
```
Seems like `mt/with-full-data-perms-for-all-users!` got removed; was it un-needed or should we add an analogue to it?

johnswanson replied:
Good call, I'll add it back in, think I missed this in a rebase.

noahmoss commented on src/metabase/models/field.clj:
```diff
@@ -166,22 +161,38 @@
+
+(defmethod mi/can-write? :model/Field
+  ([instance]
+   (if (premium-features/enable-advanced-permissions?)
+     (= :yes (data-perms/table-permission-for-user
+              api/*current-user-id*
+              :perms/manage-table-metadata
+              (field->db-id instance)
+              (:table_id instance)))
+     (mi/superuser?)))
```
I think this check should be moved to EE code. Currently `perms/data-model-write-perms-path` tries to resolve `metabase-enterprise.advanced-permissions.models.permissions/data-model-write-perms-path` and calls into it when EE functionality is enabled; otherwise it requires admin perms to write fields.

Could you do the same here, maybe with a `defenterprise` macro? We should make sure that enforcement of EE-only permissions is done in EE namespaces so that it's covered by the correct license.

noahmoss commented on src/metabase/models/field.clj:
```diff
@@ -147,14 +148,8 @@
-
-(def ^:private ^{:arglists '([table-id read-or-write])} cached-perms-object-set
-  "Cached lookup for the permissions set for a table with `table-id`. This is done so a single API call or other unit of
+(def ^:private ^{:arglists '([table-id])} cached-db-id
```
We can probably just use `metabase.models.table/table-id->database-id` for this

johnswanson committed:
use existing `table-id->database-id` function

I needed to move the function to the database model to eliminate
circular dependencies as well.

johnswanson committed:
re-add `with-full-data-perms` to test

johnswanson committed:
use defenterprise to gate advanced permissions
